src/Microsoft.AspNet.Http.Abstractions/HttpResponse.cs

```diff
@@ -24,9 +25,9 @@ public abstract class HttpResponse
 
         public abstract bool HasStarted { get; }
 
-        public abstract void OnResponseStarting(Action<object> callback, object state);
+        public abstract void OnResponseStarting(Func<object, Task> callback, object state);
```

Comment:
Do we want sync sugar for these? And while we're at it we could have sugar that doesn't require state.